src/persist-client/src/internal/apply.rs

```diff
@@ -207,6 +207,28 @@ where
             })
     }
 
+    /// WIP:
+    pub fn removable_rollups(&self) -> Vec<(SeqNo, PartialRollupKey)> {
+        let mut rollups: Vec<(SeqNo, PartialRollupKey)> =
+            self.state
+                .read_lock(&self.metrics.locks.applier_read_noncacheable, |state| {
+                    state
+                        .collections
+                        .rollups
+                        .range(..=state.seqno_since())
+                        .map(|(seqno, rollup)| (*seqno, rollup.key.to_owned()))
+                        .collect()
+                });
+
+        // as noted in StateVersions, our invariant holds that we may truncate
+        // rollups `<` than the latest rollup that is `<= seqno_since`. this
+        // means callers should not remove the latest rollup, if any, so pop it
+        // from our output:
+        rollups.pop();
```

Comment:
woah, the inclusive end range above plus this is (I think) correct but pretty subtle. I wonder if there's a way to make it more obvious

Reply:
reworked this quite a lot since you last viewed. not sure if it now hits the right level of readability, but hopefully an improvement